configure.ac

```diff
@@ -443,11 +443,32 @@ if test "x$with_dl" != "xno"; then
   AC_CHECK_LIB(dl, dlopen)
 fi
 
+AC_DEFUN([LIBATOMIC_CHECKER],[
+  AC_MSG_CHECKING(whether linking $1 -latomic works)
+  AC_LINK_IFELSE([
+    AC_LANG_SOURCE([[
+#include <atomic>
+      struct LockRequired { char data[256]; };
+      int
+      main(int, char **) {
+          return std::atomic<LockRequired>{}.is_lock_free();
+      }
```

Comment:
After reading the draft GCC libatomic ABI [specification](https://www.mail-archive.com/gcc@gcc.gnu.org/msg81382.html), I think it was a mistake to use a 256-byte atomic because there is a remote possibility that Squid does _not_ need libatomic in a given environment, libatomic is _missing_, and our new test fails even though Squid could have been built just fine. I think that our test should use the (largest) type actually used by Squid itself -- uint64_t.

AFAICT, GCC libatomic ABA requires that "the compiler shall always generate support function calls for atomic operations on the objects of [not inlineable atomic] type"[1]. Those functions are defined in GCC libatomic or equivalent. Thus, by using a definitely _not_ inlineable 256-byte atomic, we essentially require libatomic linking, which may lead to unnecessary ./configure termination.

I still suspect that std::atomic<T>::is_lock_free() _always_ requires libatomic for some unknown to me reason, even for small, inlineable T, but I do not know that, so it is best to test with what Squid is actually _using_ -- uint64_t.

```suggestion
#include <atomic>
#include <cstdint>
      int
      main(int, char **) {
          return std::atomic<uint64_t>{}.is_lock_free() ? 0 : 1;
      }
```

Why such a complex return value? I know that this test program is never executed, but I thought it would be better to return a "successful" zero result when 64-bit atomics are lock-free.

Reply:
> remote possibility that Squid does not need libatomic in a given environment, libatomic is missing, and our new test fails even though Squid could have been built just fine

To clarify: the fact that (in this scenario) both our configure tests ('with' and 'without' libatomic) fail will result in no 'ATOMICLIB' variable defined. If Squid really does not need libatomic, it should be built successfully in this case.

Reply:
> the fact that (in this scenario) both our configure tests ('with' and 'without' libatomic) fail will result in no 'ATOMICLIB' variable defined.

Ah, I thought that Amos wanted ./configure itself to fail in that case! I probably misunderstood. Thank you for correcting me.

I still think that ./configure should use for testing what we use in Squid code -- my change request stands for the already stated reasons. It is just that the implemented approach decreases the cost of a mistake, which is a good thing.

Reply:
configure failing is what I meant:

* without test succeeds - ATOMICLIB undefined
* without test fails - try with test
 * with test succeeds - ATOMICLIB defined
 * with test fails (i.e. both failed) - configure exit with AC_ERROR_MSG

Reply:
> I still think that ./configure should use ...

Done at 191084f.

> with test fails (i.e. both failed) - configure exit with AC_ERROR_MSG

Done at 4a64bd5.